src/libsync/creds/httpcredentials.cpp

```diff
@@ -412,47 +412,64 @@ bool HttpCredentials::refreshAccessToken()
-            qCWarning(lcHttpCredentials) << "Error while refreshing the token" << reply->errorString() << jsonData << jsonParseError.errorString();
-        } else if (accessToken.isEmpty()) {
-            // If the json was valid, but the reply did not contain an access token, the token
-            // is considered expired. (Usually the HTTP reply code is 400)
-            qCDebug(lcHttpCredentials) << "Expired refresh token. Logging out";
-            _refreshToken.clear();
-        } else {
-            _ready = true;
-            _password = accessToken;
-            _refreshToken = json["refresh_token"].toString();
-            persist();
-        }
-        _isRenewingOAuthToken = false;
-        for (const auto &job : _retryQueue) {
-            if (job)
-                job->retry();
+        QUrl requestTokenUrl = Utility::concatUrlPath(_account->url(), QLatin1String("/index.php/apps/oauth2/api/v1/token"));
+        if (reply->error() == QNetworkReply::NoError) {
+            auto jsonData = reply->readAll();
+            QJsonParseError jsonParseError;
+            QJsonObject json = QJsonDocument::fromJson(jsonData, &jsonParseError).object();
+            if (jsonParseError.error == QJsonParseError::NoError) {
+                QString tokenEp = json["token_endpoint"].toString();
+                if (!tokenEp.isEmpty())
+                    requestTokenUrl = tokenEp;
+            }
         }
-        _retryQueue.clear();
-        emit fetched();
+
+        QNetworkRequest req;
+        req.setHeader(QNetworkRequest::ContentTypeHeader, "application/x-www-form-urlencoded");
+
+        QString basicAuth = QString("%1:%2").arg(
+            Theme::instance()->oauthClientId(), Theme::instance()->oauthClientSecret());
+        req.setRawHeader("Authorization", "Basic " + basicAuth.toUtf8().toBase64());
+        req.setAttribute(HttpCredentials::DontAddCredentialsAttribute, true);
+
+        auto requestBody = new QBuffer;
+        QUrlQuery arguments(QString("grant_type=refresh_token&refresh_token=%1").arg(_refreshToken));
+        requestBody->setData(arguments.query(QUrl::FullyEncoded).toLatin1());
+
+        auto job = _account->sendRequest("POST", requestTokenUrl, req, requestBody);
+        job->setTimeout(qMin(30 * 1000ll, job->timeoutMsec()));
+        QObject::connect(job, &SimpleNetworkJob::finishedSignal, this, [this](QNetworkReply *reply) {
+            auto jsonData = reply->readAll();
+            QJsonParseError jsonParseError;
+            QJsonObject json = QJsonDocument::fromJson(jsonData, &jsonParseError).object();
+            QString accessToken = json["access_token"].toString();
+            if (jsonParseError.error != QJsonParseError::NoError || json.isEmpty()) {
+                // Invalid or empty JSON: Network error maybe?
+                qCWarning(lcHttpCredentials) << "Error while refreshing the token" << reply->errorString() << jsonData << jsonParseError.errorString();
```

Comment:
Is there any error state you need to set in some variable or return value?

Reply:
The idea is that the jobs get re-tried and will then process to the normal authentication method.
Yeah, maybe something better could be done, but we really don't want to disturb the user in this case anyway. This is not new code, it just got one more level of identation.